benches/metrics_bench_util/mod.rs

```diff
@@ -0,0 +1,292 @@
+//! This mod contains the benchmarks for metrics.
+//! Due to how metrics are set up, we need isolated process environments to
+//! test with metrics core on and off, so the implementation is shared but has
+//! two separate entrypoits.
+//!
+//! # Usage
+//!
+//! To run these benches, use the following commands:
+//!
+//!     cargo bench --no-default-features --features "metrics-benches" --bench metrics_on
+//!     cargo bench --no-default-features --features "metrics-benches" --bench metrics_off
+//!
+//! These will run benches with metrics core system on and off respectively.
+
+use criterion::{BatchSize, BenchmarkId, Criterion, SamplingMode, Throughput};
+use futures::{compat::Future01CompatExt, future, stream, StreamExt};
+use metrics::counter;
+use tracing::{span, Level};
+use vector::{
+    config, sinks, sources,
+    test_util::{
+        next_addr, random_lines, runtime, send_lines, start_topology, wait_for_tcp, CountReceiver,
+    },
+};
+
+#[inline]
+fn disable_metrics() {
+    std::env::set_var("DISABLE_INTERNAL_METRICS_CORE", "true");
+}
+
+#[inline]
+fn disable_metrics_tracing_integration() {
+    std::env::set_var("DISABLE_INTERNAL_METRICS_TRACING_INTEGRATION", "true");
+}
+
+#[inline]
+fn boot() {
+    vector::trace::init(false, false, "warn");
+    vector::metrics::init().expect("metrics initialization failed");
+}
+
+#[allow(dead_code)] // condition compilation
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Mode {
+    MetricsOff,
+    MetricsNoTracingIntegration,
+    MetricsOn,
+}
+
+impl Mode {
+    fn as_str(&self) -> &'static str {
+        match self {
+            Mode::MetricsOff => "metrics_off",
+            Mode::MetricsNoTracingIntegration => "metrics_no_tracing_integration",
+            Mode::MetricsOn => "metrics_on",
+        }
+    }
+}
+
+/// Due to the nature of how metrics and tracing systems are set up, we need
+/// separate process spaces to measure with and without them being enabled.
+#[inline]
```

Comment:
If you want the instructions in the body of the function to just be compiled "inline" in the asm, rather than just being at some dedicated section and invoked via `call` instruction. I.e. use to hint compiler that the function body is probably small enough that it doesn't worth the overhead of doing a `call` to jump to it and back. The cost is more instructions (thus a bit of extra memory in the binary - and thus in the ram as well) in each of the call sites. The benefit is extra locality.

Reply:
:thinking: yeah, I see. Do you just sprinkle it on functions that look small enough to you then? Or are there other heuristics?

Reply:
In this case, the deciding decision is that this `fn` is expected to be used as if its body was just written at `main`. I mean, that's the idea, the design. Each of our bench binaries is expected to have no more than one (exactly one to be more precise) of this call, so it won't cause any problem. Quite frankly, the compiler would most likely deduce it on its own - and inline it anyway - but since I recognized this is the behaviour I want, I added the hint at the source level.

In general, it's not the size of the fn per se, but how you imagine you'd like the generated machine code to look like you have to think about. Quite often there's no need to (or you even can't) be very precise though, as the compiler knows better - but you can at least hint it your intentions.